File: 202-happy-number/202-happy-number.cpp
class Solution {
public:
    bool isHappy(int n) {
        int cur = n;
        unordered_map<int, int> seen;
        
        while (true) {
            if (cur == 1) {
                return true;
            }
            
            ++seen[cur];
            if (seen[cur] >= 2) {
                return false;
            }
            
            int temp = 0;
            while (cur) {
                temp += (cur % 10) * (cur % 10);
                cur /= 10;
            }
            cur = temp;
        }
    }
};

File: 202-happy-number/202-happy-number.py
class Solution:
    def isHappy(self, n: int) -> bool:
        seen = []
        
        def next(n):
            x = sum(int(dig)**2 for dig in str(n))
            if x == 1:
                return True
            elif x in seen:
                return False
            else:
                seen.append(x)
                return next(x)
        
        return next(n)